src/Shared/ChangeWaves.cs

```diff
@@ -88,120 +105,74 @@ internal static ChangeWaveConversionState ConversionState
+            else if (_cachedWave < LowestWave)
             {
                 ConversionState = ChangeWaveConversionState.OutOfRotation;
-                DisabledWave = LowestWave;
+                _cachedWave = LowestWave;
                 return;
             }
-            else if (changeWave > HighestWaveAsVersion)
+            else if (_cachedWave > HighestWave)
             {
                 ConversionState = ChangeWaveConversionState.OutOfRotation;
-                DisabledWave = HighestWave;
+                _cachedWave = HighestWave;
                 return;
             }
 
-            // Ensure it's set to an existing version within the current rotation
-            if (!AllWavesAsVersion.Contains(changeWave))
-            {
-                foreach (Version wave in AllWavesAsVersion)
-                {
-                    if (wave > changeWave)
-                    {
-                        ConversionState = ChangeWaveConversionState.Valid;
-                        DisabledWave = wave.ToString();
-                        return;
-                    }
-                }
-            }
-
+            // _cachedWave is somewhere between valid waves, find the next valid version.
+            _cachedWave = AllWaves.First((x) => x > _cachedWave);
             ConversionState = ChangeWaveConversionState.Valid;
-            DisabledWave = changeWave.ToString();
-        }
-
-        /// <summary>
-        /// Compares the passed wave to the MSBuildDisableFeaturesFromVersion environment variable.
-        /// Version MUST be of the format: "xx.yy".
-        /// </summary>
-        /// <param name="wave">The version to compare.</param>
-        /// <returns>A bool indicating whether the feature behind a version is enabled.</returns>
-        public static bool AreFeaturesEnabled(string wave)
-        {
-            Version waveToCheck;
-
-            // When a caller passes an invalid wave, fail the build.
-            ErrorUtilities.VerifyThrow(Version.TryParse(wave.ToString(), out waveToCheck),
-                                       $"Argument 'wave' passed with invalid format." +
-                                       $"Please use pre-existing const strings or define one with format 'xx.yy");
-
-            return AreFeaturesEnabled(waveToCheck);
         }
 
         /// <summary>
-        /// Compares the passed wave to the MSBuildDisableFeaturesFromVersion environment variable.
+        /// Determines whether features behind the given wave are enabled.
         /// </summary>
         /// <param name="wave">The version to compare.</param>
-        /// <returns>A bool indicating whether the version is enabled.</returns>
+        /// <returns>A bool indicating whether the change wave is enabled.</returns>
         public static bool AreFeaturesEnabled(Version wave)
         {
-            if (_state == ChangeWaveConversionState.NotConvertedYet)
+            if (ShouldApplyChangeWave)
```

Comment:
This check is covered in ApplyChangeWave. Also, is there any time ApplyChangeWave could miss being called from Evaluator or that AreFeaturesEnabled could be called earlier than ApplyChangeWave in Evaluate? Wondering if you can get rid of this entirely.

Reply:
Good point, I can get rid of this check. I can't get rid of the call to `ApplyChangeWave` in this function though.

> lso, is there any time ApplyChangeWave could miss being called from Evaluator or that AreFeaturesEnabled could be called earlier than ApplyChangeWave in Evaluate?

Yes, it doesn't miss being called but `AreFeaturesEnabled` (and therefore `ApplyChangeWave`) is first called from [XMake](https://github.com/dotnet/msbuild/blob/master/src/MSBuild/XMake.cs#L1733).